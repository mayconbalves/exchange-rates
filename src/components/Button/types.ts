export type ButtonProps = {
  disabled?: boolean
  onClick: () => void
  title: string
}
