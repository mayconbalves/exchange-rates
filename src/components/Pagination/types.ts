export type PaginationProps = {
  currentPage: number
  itemsPerPage: number
  totalItems: number
  onPageChange: (page: number) => void
}
