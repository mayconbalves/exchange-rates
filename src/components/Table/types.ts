export type TableDate = {
  rate: string
  value: number
  date: string
  timestamp: number
}

export type SortConfig = {
  key: keyof TableDate | null
  direction: 'ascending' | 'descending'
}

export type TableProps = {
  rates: TableDate[]
  date: string
}
