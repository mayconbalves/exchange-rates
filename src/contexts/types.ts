export type ToastContextProps = {
  showToast: (message: string, type: 'success' | 'error') => void
}
