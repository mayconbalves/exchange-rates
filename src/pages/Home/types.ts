export type Registrations = {
  id: string
  admissionDate: string
  email: string
  employeeName: string
  status: 'REPROVED' | 'APPROVED' | 'REVIEW'
  cpf: string
}
