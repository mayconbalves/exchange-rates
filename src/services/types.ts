export type RateItem = {
  date: string
  rate: string
  value: number
  timestamp: number
}

export type ApiResponse = {
  success: boolean
  timestamp: number
  base: string
  date: string
  rates: Record<string, number>
}
