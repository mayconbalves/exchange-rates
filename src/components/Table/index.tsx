import React, { useMemo, useState } from 'react'
import { FaSortDown, FaSortUp } from 'react-icons/fa'
import styles from './styles.module.css'
import { SortConfig, TableDate, TableProps } from './types'

const Table: React.FC<TableProps> = ({ rates, date }) => {
  const [sortConfig, setSortConfig] = useState<SortConfig>({
    key: null,
    direction: 'ascending'
  })

  const formatTimestamp = (timestamp: number) => {
    const date = new Date(timestamp * 1000)
    const formattedDate = date.toLocaleString('pt-BR', {
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit'
    })

    return formattedDate
  }

  const sortedData = useMemo(() => {
    const validData = Array.isArray(rates) ? rates : []
    const sortableItems = [...validData]
    if (sortConfig.key !== null) {
      sortableItems.sort((a, b) => {
        if (a[sortConfig.key!] < b[sortConfig.key!])
          return sortConfig.direction === 'ascending' ? -1 : 1
        if (a[sortConfig.key!] > b[sortConfig.key!])
          return sortConfig.direction === 'ascending' ? 1 : -1
        return 0
      })
    }
    return sortableItems
  }, [rates, sortConfig])

  const handleSort = (key: keyof TableDate, direction: 'ascending' | 'descending') => {
    setSortConfig({ key, direction })
  }

  const getIconColor = (key: keyof TableDate, direction: 'ascending' | 'descending') => {
    return sortConfig.key === key && sortConfig.direction === direction ? '#3498db' : '#bbb'
  }

  return (
    <table className={styles.table}>
      <thead>
        <tr>
          <th className={styles.header}>
            <div className={styles.header__content}>
              Moeda
              <div className={styles.icon__container}>
                <FaSortUp
                  style={{
                    color: getIconColor('rate', 'ascending'),
                    cursor:
                      sortConfig.key === 'rate' && sortConfig.direction === 'ascending'
                        ? 'not-allowed'
                        : 'pointer'
                  }}
                  onClick={() =>
                    sortConfig.key !== 'rate' || sortConfig.direction !== 'ascending'
                      ? handleSort('rate', 'ascending')
                      : undefined
                  }
                />
                <FaSortDown
                  style={{
                    color: getIconColor('rate', 'descending'),
                    cursor:
                      sortConfig.key === 'rate' && sortConfig.direction === 'descending'
                        ? 'not-allowed'
                        : 'pointer'
                  }}
                  onClick={() =>
                    sortConfig.key !== 'rate' || sortConfig.direction !== 'descending'
                      ? handleSort('rate', 'descending')
                      : undefined
                  }
                />
              </div>
            </div>
          </th>
          <th className={styles.header}>
            <div className={styles.header__content}>
              Valor (em EUR)
              <div className={styles.icon__container}>
                <FaSortUp
                  style={{
                    color: getIconColor('value', 'ascending'),
                    cursor:
                      sortConfig.key === 'value' && sortConfig.direction === 'ascending'
                        ? 'not-allowed'
                        : 'pointer'
                  }}
                  onClick={() =>
                    sortConfig.key !== 'value' || sortConfig.direction !== 'ascending'
                      ? handleSort('value', 'ascending')
                      : undefined
                  }
                />
                <FaSortDown
                  style={{
                    color: getIconColor('value', 'descending'),
                    cursor:
                      sortConfig.key === 'value' && sortConfig.direction === 'descending'
                        ? 'not-allowed'
                        : 'pointer'
                  }}
                  onClick={() =>
                    sortConfig.key !== 'value' || sortConfig.direction !== 'descending'
                      ? handleSort('value', 'descending')
                      : undefined
                  }
                />
              </div>
            </div>
          </th>
          <th className={styles.header}>
            <div className={styles.header__content}>
              Horário
              <div className={styles.icon__container}>
                <FaSortUp
                  style={{
                    color: getIconColor('date', 'ascending'),
                    cursor:
                      sortConfig.key === 'date' && sortConfig.direction === 'ascending'
                        ? 'not-allowed'
                        : 'pointer'
                  }}
                  onClick={() =>
                    sortConfig.key !== 'date' || sortConfig.direction !== 'ascending'
                      ? handleSort('date', 'ascending')
                      : undefined
                  }
                />
                <FaSortDown
                  style={{
                    color: getIconColor('date', 'descending'),
                    cursor:
                      sortConfig.key === 'date' && sortConfig.direction === 'descending'
                        ? 'not-allowed'
                        : 'pointer'
                  }}
                  onClick={() =>
                    sortConfig.key !== 'date' || sortConfig.direction !== 'descending'
                      ? handleSort('date', 'descending')
                      : undefined
                  }
                />
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        {sortedData.map((item, index) => (
          <tr key={index} className={styles.row}>
            <td className={styles.cell}>
              <p>{item.rate}</p>
            </td>
            <td className={styles.cell}>
              <p>{item.value.toFixed(2)}</p>
            </td>
            <td className={styles.cell}>
              <p>{`${date} às ${formatTimestamp(item.timestamp)}`}</p>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}

export default Table
