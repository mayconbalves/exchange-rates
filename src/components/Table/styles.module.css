.table {
  border-collapse: collapse;
  margin: 20px 0;
  font-size: 1rem;
  text-align: left;
  border: 2px solid #000000;
  overflow: hidden;
}

.header {
  padding: 12px;
  background-color: #f1f1f1;
  font-weight: bold;
  color: #333;
  border: 2px solid #000000;
  cursor: pointer;
  min-width: 200px;
  position: relative;
}

.header__content {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.icon__container {
  display: flex;
  flex-direction: column;
  margin-left: 8px;
}

.row {
  border: 2px solid #000000;
}

.cell {
  padding: 12px;
  color: #555;
  border: 2px solid #000000;

  p {
    border: 2px solid #000000;
    padding: 12px;
  }
}
