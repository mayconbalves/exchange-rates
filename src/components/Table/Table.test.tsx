import { render, screen } from '@testing-library/react'
import Table from '.'
import { TableProps } from './types'

describe('Table Component', () => {
  const sampleRates = [
    { rate: 'USD', value: 1.12, date: '10:00', timestamp: 1731285422 },
    { rate: 'GBP', value: 0.85, date: '10:05', timestamp: 1731285422 },
    { rate: 'JPY', value: 130.5, date: '10:10', timestamp: 1731285422 }
  ]

  const defaultProps: TableProps = {
    rates: sampleRates,
    date: '2024-11-09'
  }

  it('Should render the table with correct headers', () => {
    render(<Table {...defaultProps} />)
    expect(screen.getByText(/Moeda/i)).toBeInTheDocument()
    expect(screen.getByText(/Valor \(em EUR\)/i)).toBeInTheDocument()
    expect(screen.getByText(/HorÃ¡rio/i)).toBeInTheDocument()
  })

  it('Should render correct number of rows based on rates', () => {
    render(<Table {...defaultProps} />)
    const rows = screen.getAllByRole('row')
    expect(rows.length).toBe(sampleRates.length + 1)
  })

  it('Should be create snapshot', () => {
    const { container } = render(<Table {...defaultProps} />)

    expect(container).toMatchSnapshot()
  })
})
