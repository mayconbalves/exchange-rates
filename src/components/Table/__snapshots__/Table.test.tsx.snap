// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component Should be create snapshot 1`] = `
<div>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th
          class="header"
        >
          <div
            class="header__content"
          >
            Moeda
            <div
              class="icon__container"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(187, 187, 187); cursor: pointer;"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M279 224H41c-21.4 0-32.1-25.9-17-41L143 64c9.4-9.4 24.6-9.4 33.9 0l119 119c15.2 15.1 4.5 41-16.9 41z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(187, 187, 187); cursor: pointer;"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
                />
              </svg>
            </div>
          </div>
        </th>
        <th
          class="header"
        >
          <div
            class="header__content"
          >
            Valor (em EUR)
            <div
              class="icon__container"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(187, 187, 187); cursor: pointer;"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M279 224H41c-21.4 0-32.1-25.9-17-41L143 64c9.4-9.4 24.6-9.4 33.9 0l119 119c15.2 15.1 4.5 41-16.9 41z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(187, 187, 187); cursor: pointer;"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
                />
              </svg>
            </div>
          </div>
        </th>
        <th
          class="header"
        >
          <div
            class="header__content"
          >
            Horário
            <div
              class="icon__container"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(187, 187, 187); cursor: pointer;"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M279 224H41c-21.4 0-32.1-25.9-17-41L143 64c9.4-9.4 24.6-9.4 33.9 0l119 119c15.2 15.1 4.5 41-16.9 41z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(187, 187, 187); cursor: pointer;"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
                />
              </svg>
            </div>
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="row"
      >
        <td
          class="cell"
        >
          <p>
            USD
          </p>
        </td>
        <td
          class="cell"
        >
          <p>
            1.12
          </p>
        </td>
        <td
          class="cell"
        >
          <p>
            2024-11-09 às 21:37:02
          </p>
        </td>
      </tr>
      <tr
        class="row"
      >
        <td
          class="cell"
        >
          <p>
            GBP
          </p>
        </td>
        <td
          class="cell"
        >
          <p>
            0.85
          </p>
        </td>
        <td
          class="cell"
        >
          <p>
            2024-11-09 às 21:37:02
          </p>
        </td>
      </tr>
      <tr
        class="row"
      >
        <td
          class="cell"
        >
          <p>
            JPY
          </p>
        </td>
        <td
          class="cell"
        >
          <p>
            130.50
          </p>
        </td>
        <td
          class="cell"
        >
          <p>
            2024-11-09 às 21:37:02
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
